sdk/java/pulumi/src/main/java/com/pulumi/serialization/internal/ResourcePackages.java

```diff
@@ -1,152 +1,199 @@
                             ));
                         }
-                    })
-                    .filter(c -> !Reflection.isNestedClass(c))
-                    .map(c -> {
-                        //noinspection unchecked
-                        return (Class<Resource>) c; // checked in filter above
-                    })
-                    .map(c -> Tuples.of(c.getAnnotation(ResourceType.class), c))
-                    .collect(
-                            collectingAndThen(
-                                    Collectors.groupingBy(
-                                            g -> g.t1.type(),
-                                            Collectors.mapping(
-                                                    g -> Tuples.of(Optionals.ofBlank(g.t1.version()), g.t2),
-                                                    ImmutableList.<Tuple2<Optional<String>, Class<Resource>>>toImmutableList()
-                                            )
-                                    ),
-                                    ImmutableMap::copyOf
-                            )
-                    );
+                        return true;
+                    } else {
+                        throw new IllegalStateException(String.format(
+                                "Expected class '%s' to be annotated with '%s', to be assignable from '%s', but it was not.",
+                                c.getTypeName(), ResourceType.class.getTypeName(), Resource.class.getTypeName()
+                        ));
+                    }
+                })
+                .filter(c -> !Reflection.isNestedClass(c))
+                .map(c -> {
+                    //noinspection unchecked
+                    return (Class<Resource>) c; // checked in filters above
+                })
+                .map(c -> new ResourceTypeAndClass(c.getAnnotation(ResourceType.class), c))
+                .collect(
+                        collectingAndThen(
+                                groupingBy(
+                                        g -> g.annotation.type(),
+                                        Collectors.mapping(
+                                                g -> new ResourceTypeEntry(Optionals.ofBlank(g.annotation.version()), g.type),
+                                                ImmutableList.<ResourceTypeEntry>toImmutableList()
+                                        )
+                                ),
+                                ImmutableMap::copyOf
+                        )
+                );
+    }
 
-        } catch (IOException e) {
-            throw new IllegalStateException(String.format("Failed to read class path: %s", e.getMessage()), e);
-        }
+    @SuppressWarnings("UnstableApiUsage")
+    private static boolean excludePackages(ClassInfo c) {
+        return !c.getPackageName().isBlank()
+                && !c.getPackageName().startsWith("META-INF")
```

Comment:
this list is not ideal, but should be good enough until we detect performance issues